### ISSUE TITLE
Aggregated events can outlive the events they duplicate

### ISSUE THREAD

**Issue by kensimon** [kubernetes/kubernetes#45971]
**Is this a request for help?**: No

**What keywords did you search in Kubernetes issues before filing this one?** : is:issue is:open in:title event 

Original report: #45818

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug report

**Kubernetes version** (use `kubectl version`):

Latest master, although same bug is in released versions

```
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.3.625+b353700bfc1ee9-dirty", GitCommit:"b353700bfc1ee95b8f9f827c9cc818cf2ab3f943", GitTreeState:"dirty", BuildDate:"2017-05-17T15:22:37Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.3.625+b353700bfc1ee9-dirty", GitCommit:"b353700bfc1ee95b8f9f827c9cc818cf2ab3f943", GitTreeState:"dirty", BuildDate:"2017-05-17T15:16:24Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: local, with `hack/local-up-cluster.sh`
- **OS** (e.g. from /etc/os-release): Ubuntu 17.04
- **Kernel** (e.g. `uname -a`): 4.10.0-20-generic
- **Install tools**:
- **Others**:

**What happened**:

Events are currently correlated in a strange way: if too many identical events are produced, a *new* event is created with a message of `(events with common reason combined)`

From the [design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/event_compression.md), this was intentional:

> If the EventAggregator observes a similar event produced 10 times in a 10 minute window, it drops the event that was provided as input and creates a new event that differs only on the message. The message denotes that this event is used to group similar events that matched on reason. This aggregated Event is then used in the event processing sequence.

It's possible that event TTL's weren't considered when this decision was made, because the original event may drop off and expire, leaving the new event to be continually bumped/modified with its `count` incrementing.  This isn't just a temporary scenario, the event that keeps getting bumped can effectively stay around forever because it keeps getting updated, whereas the original event that contains the actual message text will eventually expire.

This can be easily reproduced by launching kube-apiserver with a low TTL (like `--event-ttl=1m0s`) and launching a crashing pod.  The real reason for the crashing eventually goes away:

```
  FirstSeen	LastSeen	Count	From			SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----			-------------			--------	------		-------
...
  59s		59s		1	kubelet, 127.0.0.1					Warning		FailedSync	Error syncing pod, skipping: failed to "StartContainer" for "lifecycler" with rpc error: code = 2 desc = failed to start container "4669ec0a8e1fa478466466b5b161665204472293991aa47b36604f480bd7e16d": Error response from daemon: {"message":"oci runtime error: container_linux.go:247: starting container process caused \"exec: \\\"doesnotexist\\\": executable file not found in $PATH\"\n"}: "Start Container Failed"
```

turns into:

```
  FirstSeen	LastSeen	Count	From			SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----			-------------			--------	------		-------
...
  36m		4s		279	kubelet, 127.0.0.1					Warning		FailedSync	(events with common reason combined)
```

with the original event gone.

The original PR that brought that behavior in is here: https://github.com/kubernetes/kubernetes/pull/16798

Question I have for @derekwaynecarr or anyone else who can shed light here: Why is a new event created as an aggregation point, rather than just bumping the count and updating the message of the original one?  I don't want to pull too far on this thread here because I think there's some broad agreement that events need an overhaul, but I was wondering if you had thoughts here.

**Comment by kensimon** [comment]
cc @derekwaynecarr

**Comment by derekwaynecarr** [comment]
At the time aggregation was added, we were bleeding from event spam killing etcd.  We had many events that varied only slightly on message because they would insert things like the pod name.  For example, we had controllers that would create a pod be denied, and would throw an event "Unable to create pod pod-name" and pod-name would change every time.  As a result, we just dropped the message, and bumped the reason.  If we look to tweak anything, it needs to still protect us against spam produced by subtle variants in message strings.
